// track.js/lib/integrations/adroll.js
// Generated by CoffeeScript 1.11.1
var AdRoll, Integration, identity, page,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

Integration = require('../integration');

identity = require('../identity');

page = require('../page');

module.exports = AdRoll = (function(superClass) {
  extend(AdRoll, superClass);

  AdRoll.prototype.src = {
    type: 'script',
    url: {
      https: 'https://s.adroll.com/j/roundtrip.js',
      http: 'https://a.adroll.com/j/roundtrip.js'
    }
  };

  function AdRoll(opts) {
    this.opts = opts;
  }

  AdRoll.prototype.init = function() {
    if (window.__adroll_loaded != null) {
      return;
    }
    window.adroll_adv_id = this.opts.advId;
    window.adroll_pix_id = this.opts.pixId;
    return window.__adroll_loaded = true;
  };

  AdRoll.prototype.identify = function(userId, props, cb) {
    var ref;
    return window.adroll_email((ref = props.email) != null ? ref : userId);
  };

  AdRoll.prototype.page = function(category, name, props, cb) {
    this.track(page.title());
    return cb(null);
  };

  AdRoll.prototype.track = function(event, props, cb) {
    var data, ref;
    data = {
      user_id: identity.id(),
      order_id: props.orderId,
      adroll_conversion_value: props.revenue,
      adroll_currency: (ref = props.currency) != null ? ref : 'USD'
    };
    return window.__adroll.record_user(data);
  };

  return AdRoll;

})(Integration);
