// track.js/lib/integrations/bing.js
// Generated by CoffeeScript 1.11.1
var Bing, Integration,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

Integration = require('../integration');

module.exports = Bing = (function(superClass) {
  extend(Bing, superClass);

  Bing.prototype.src = {
    type: 'script',
    url: '//bat.bing.com/bat.js'
  };

  function Bing(opts) {
    this.opts = opts;
  }

  Bing.prototype.init = function() {
    window.uetq = window.uetq || [];
    return this.load(function() {
      var setup;
      setup = {
        ti: self.options.tagId,
        q: window.uetq
      };
      window.uetq = new window.UET(setup);
      return this.ready();
    });
  };

  Bing.prototype.page = function(event, props, cb) {
    window.uetq.push('pageLoad');
    return cb(null);
  };

  Bing.prototype.track = function(event, props, cb) {
    window.heap.track(event, props);
    cb(null);
    event = {
      ea: 'track',
      el: event
    };
    if (event.ec == null) {
      event.ec = props.category;
    }
    if (event.gv == null) {
      event.gv = props.revenue;
    }
    return window.uetq.push(event);
  };

  return Bing;

})(Integration);
