// track.js/lib/integrations/custom.js
// Generated by CoffeeScript 1.11.1
var Custom, Integration,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

Integration = require('../integration');

module.exports = Custom = (function(superClass) {
  extend(Custom, superClass);

  function Custom(opts) {
    this.opts = opts;
    if (this.src == null) {
      this.src = this.opts.src;
    }
    this.fn = new Function(this.opts.code);
  }

  Custom.prototype.track = function(event, props, cb) {
    var err;
    if (event !== this.opts.event) {
      return;
    }
    try {
      return this.fn(event, props, cb);
    } catch (error) {
      err = error;
      return this.log("Custom integration failed, " + (err.toString()));
    }
  };

  return Custom;

})(Integration);
