// track.js/lib/integrations/google-adwords.js
// Generated by CoffeeScript 1.11.1
var GoogleAdWords, Integration,
  extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
  hasProp = {}.hasOwnProperty;

Integration = require('../integration');

module.exports = GoogleAdWords = (function(superClass) {
  extend(GoogleAdWords, superClass);

  GoogleAdWords.prototype.src = {
    type: 'script',
    url: '//www.googleadservices.com/pagead/conversion_async.js'
  };

  function GoogleAdWords(opts) {
    this.opts = opts;
    this.queue = [];
  }

  GoogleAdWords.prototype.init = function() {
    if (window.google_trackConversion != null) {
      return;
    }
    return window.google_trackConversion = (function(_this) {
      return function(event) {
        return _this.queue.push(event);
      };
    })(this);
  };

  GoogleAdWords.prototype.load = function(cb) {
    return GoogleAdWords.__super__.load.call(this, (function(_this) {
      return function() {
        var event, i, len, ref;
        if (_this.queue.length) {
          ref = _this.queue;
          for (i = 0, len = ref.length; i < len; i++) {
            event = ref[i];
            google_trackConversion(event);
          }
        }
        return cb(null);
      };
    })(this));
  };

  GoogleAdWords.prototype.page = function(category, name, props, cb) {
    if (arguments.length === 1) {
      name = category;
    }
    if (name !== this.opts.event) {
      return;
    }
    google_trackConversion({
      google_conversion_id: this.opts.id,
      google_custom_params: props,
      google_remarketing_only: true
    });
    return cb(null);
  };

  GoogleAdWords.prototype.track = function(event, props, cb) {
    if (event !== this.opts.event) {
      return;
    }
    google_trackConversion({
      google_conversion_id: this.opts.id,
      google_custom_params: props,
      google_conversion_language: 'en',
      google_conversion_format: '3',
      google_conversion_color: 'ffffff',
      google_conversion_label: event,
      google_conversion_value: props.total,
      google_remarketing_only: false
    });
    return cb(null);
  };

  return GoogleAdWords;

})(Integration);
